src/Azure.Functions.Cli/Common/AuthManager.cs

```diff
@@ -0,0 +1,405 @@
+
+        }
+
+        private List<requiredResourceAccess> GetRequiredResourceAccesses()
+        {
+            var bindings = ExtensionsHelper.GetBindingsWithDirection();
+
+            // Required for basic Easy Auth / Middleware authentication
+            var resourceList = new List<requiredResourceAccess>
+            {
+                new requiredResourceAccess
+                {
+                    resourceAppId = AADConstants.ServicePrincipals.AzureADGraph,
+                    resourceAccess = new resourceAccess[]
+                    {
+                        new resourceAccess
+                        {
+                            type = AADConstants.ResourceAccessTypes.User,
+                            id = AADConstants.Permissions.EnableSSO.ToString()
+                        }
+                    }
+                },
+            };
+
+            // Determine which Microsoft Graph permissions are necessary,
+            // Based on which I/O bindings are in the user's functions
+            HashSet<Guid> requiredPermissions = new HashSet<Guid>();
+
+            foreach (var binding in bindings)
+            {
+                // Determine the required permissions for this binding
+                if (AADConstants.PermissionMap.ContainsKey(binding))
+                {
+                    requiredPermissions.UnionWith(AADConstants.PermissionMap[binding]);
+                }
+            }
+
+            var resourceAccessList = new List<resourceAccess>();
+            foreach (var permission in requiredPermissions)
+            {
+                resourceAccessList.Add(new resourceAccess
+                {
+                    type = AADConstants.ResourceAccessTypes.User,
+                    id = permission.ToString()
+                });
+            }
+
+            if (resourceAccessList.Count > 0)
+            {
+                resourceList.Add(new requiredResourceAccess
+                {
+                    resourceAppId = AADConstants.ServicePrincipals.MicrosoftGraph,
+                    resourceAccess = resourceAccessList.ToArray()
+                });
+            }
+
+            return resourceList;
+        }
+
+        private void CreateAndCommitAuthSettings(string hostName, string clientId, string clientSecret, string tenant, List<string> replyUrls)
+        {
+            // The WEBSITE_AUTH_ALLOWED_AUDIENCES setting is of the form "{replyURL1} {replyURL2}"
+            string serializedReplyUrls = string.Join(" ", replyUrls);
+
+            // Create a local auth .json file that will be used by the middleware
+            var middlewareAuthSettingsFile = SecretsManager.MiddlewareAuthSettingsFileName;
+            MiddlewareAuthSettings = new AuthSettingsFile(middlewareAuthSettingsFile);
+            MiddlewareAuthSettings.SetAuthSetting("WEBSITE_AUTH_ALLOWED_AUDIENCES", serializedReplyUrls);
```

Comment:
GitHub isn't letting me reply to our existing comment thread on this, but the portal behavior you observed is wrong, and a bug which can cause CRIs. The allowed_audiences should NOT contain /.auth/login/aad/callback.